File: LeetCode/Medium/0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        a = dict()
        for str in strs:
            key = tuple(sorted(str))
            if key in a:
                a[key].append(str)
            else:
                a[key] = [str]
        return a.values()

File: LeetCode/Medium/0049-group-anagrams/0049-group-anagrams.kt
class Solution {

    val words = HashMap<String, ArrayList<String>>()

    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        for (word in strs) {
            val key = generateKey(word)

            if (key !in words) {
                words[key] = arrayListOf()
            }

            words[key]!! += word
        }

        return words.values.toList()
    }

    inline fun generateKey(word: String): String {
        val freq = CharArray(26)

        for (c in word) {
            freq[c - 'a']++
        }

        return freq.concatToString()
    }
}
